example_configs/zookeeper.yaml

```diff
@@ -2,22 +2,28 @@ rules:
   # replicated Zookeeper
   - pattern: "org.apache.ZooKeeperService<name0=ReplicatedServer_id(\\d+)><>(\\w+)"
     name: "zookeeper_$2"
+    type: GAUGE
```

Comment:
@brian-brazil For example this pattern produce:
```
# HELP zookeeper_quorumsize QuorumSize (org.apache.ZooKeeperService<name0=ReplicatedServer_id1><>QuorumSize)
# TYPE zookeeper_quorumsize untyped
zookeeper_quorumsize 3.0
```
I'm sure it is a guage.

Reply:
What about everything else matching this rule?

Reply:
```
# HELP zookeeper_portunification PortUnification (org.apache.ZooKeeperService<name0=ReplicatedServer_id1><>PortUnification)
# TYPE zookeeper_portunification untyped
zookeeper_portunification 0.0
# HELP zookeeper_quorumsize QuorumSize (org.apache.ZooKeeperService<name0=ReplicatedServer_id1><>QuorumSize)
# TYPE zookeeper_quorumsize untyped
zookeeper_quorumsize 3.0
# HELP zookeeper_sslquorum SslQuorum (org.apache.ZooKeeperService<name0=ReplicatedServer_id1><>SslQuorum)
# TYPE zookeeper_sslquorum untyped
zookeeper_sslquorum 0.0
```
It's a full list of metrics

Reply:
For above rule.

Reply:
Those all look like gauges. What about the rest?

Reply:
Should we examine the type rule by rule? I went through the all rules and all of them look like gauges.

Reply:
Yes, but if they all look like gauges if you can dump all of the /metrics and I can confirm quickly.

Reply:
`- pattern: "org.apache.ZooKeeperService<name0=ReplicatedServer_id(\\d+), name1=replica.(\\d+)><>(\\w+)"`
Dump:
```
# HELP zookeeper_minsessiontimeout MinSessionTimeout (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1><>MinSessionTimeout)
# TYPE zookeeper_minsessiontimeout gauge
zookeeper_minsessiontimeout{replicaId="1",} 4000.0
# HELP zookeeper_configversion ConfigVersion (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1><>ConfigVersion)
# TYPE zookeeper_configversion gauge
zookeeper_configversion{replicaId="1",} 0.0
# HELP zookeeper_initlimit InitLimit (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1><>InitLimit)
# TYPE zookeeper_initlimit gauge
zookeeper_initlimit{replicaId="1",} 5.0
# HELP zookeeper_maxsessiontimeout MaxSessionTimeout (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1><>MaxSessionTimeout)
# TYPE zookeeper_maxsessiontimeout gauge
zookeeper_maxsessiontimeout{replicaId="1",} 40000.0
# HELP zookeeper_leader Leader (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.2><>Leader)
# TYPE zookeeper_leader gauge
zookeeper_leader{replicaId="2",} 0.0
zookeeper_leader{replicaId="1",} 0.0
zookeeper_leader{replicaId="3",} 1.0
# HELP zookeeper_partofensemble PartOfEnsemble (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1><>PartOfEnsemble)
# TYPE zookeeper_partofensemble gauge
zookeeper_partofensemble{replicaId="1",} 1.0
# HELP zookeeper_electiontype ElectionType (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1><>ElectionType)
# TYPE zookeeper_electiontype gauge
zookeeper_electiontype{replicaId="1",} 3.0
# HELP zookeeper_tick Tick (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1><>Tick)
# TYPE zookeeper_tick gauge
zookeeper_tick{replicaId="1",} 0.0
# HELP zookeeper_ticktime TickTime (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1><>TickTime)
# TYPE zookeeper_ticktime gauge
zookeeper_ticktime{replicaId="1",} 2000.0
# HELP zookeeper_maxclientcnxnsperhost MaxClientCnxnsPerHost (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1><>MaxClientCnxnsPerHost)
# TYPE zookeeper_maxclientcnxnsperhost gauge
zookeeper_maxclientcnxnsperhost{replicaId="1",} 60.0
# HELP zookeeper_synclimit SyncLimit (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1><>SyncLimit)
# TYPE zookeeper_synclimit gauge
zookeeper_synclimit{replicaId="1",} 2.0
```

Reply:
zookeeper_packetssent, zookeeper_txnlogelapsedsynctime and zookeeper_packetsreceived sounds like they might be counters. Do you know what exactly they mean?

Reply:
Nope. They are from `- pattern: "org.apache.ZooKeeperService<name0=ReplicatedServer_id(\\d+), name1=replica.(\\d+), name2=(\\w+)><>(\\w+)"` as well as other metrics. It better to leave untyped than.

Reply:
If we're doing types, we should have it right for everything. Can you look into what types those are?

Reply:
This is a dump of `- pattern: "org.apache.ZooKeeperService<name0=ReplicatedServer_id(\\d+), name1=replica.(\\d+), name2=(\\w+)><>(\\w+)"` rule.
```
# HELP zookeeper_avgrequestlatency AvgRequestLatency (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1, name2=Follower><>AvgRequestLatency)
# TYPE zookeeper_avgrequestlatency untyped
zookeeper_avgrequestlatency{memberType="Follower",replicaId="1",} 0.0
# HELP zookeeper_maxrequestlatency MaxRequestLatency (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1, name2=Follower><>MaxRequestLatency)
# TYPE zookeeper_maxrequestlatency untyped
zookeeper_maxrequestlatency{memberType="Follower",replicaId="1",} 0.0
# HELP zookeeper_maxsessiontimeout MaxSessionTimeout (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1, name2=Follower><>MaxSessionTimeout)
# TYPE zookeeper_maxsessiontimeout untyped
zookeeper_maxsessiontimeout{memberType="Follower",replicaId="1",} 40000.0
# HELP zookeeper_datadirsize DataDirSize (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1, name2=Follower><>DataDirSize)
# TYPE zookeeper_datadirsize untyped
zookeeper_datadirsize{memberType="Follower",replicaId="1",} 8.05470348E8
# HELP zookeeper_jutemaxbuffersize JuteMaxBufferSize (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1, name2=Follower><>JuteMaxBufferSize)
# TYPE zookeeper_jutemaxbuffersize untyped
zookeeper_jutemaxbuffersize{memberType="Follower",replicaId="1",} 1048575.0
# HELP zookeeper_outstandingrequests OutstandingRequests (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1, name2=Follower><>OutstandingRequests)
# TYPE zookeeper_outstandingrequests untyped
zookeeper_outstandingrequests{memberType="Follower",replicaId="1",} 0.0
# HELP zookeeper_minrequestlatency MinRequestLatency (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1, name2=Follower><>MinRequestLatency)
# TYPE zookeeper_minrequestlatency untyped
zookeeper_minrequestlatency{memberType="Follower",replicaId="1",} 0.0
# HELP zookeeper_packetsreceived PacketsReceived (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1, name2=Follower><>PacketsReceived)
# TYPE zookeeper_packetsreceived untyped
zookeeper_packetsreceived{memberType="Follower",replicaId="1",} 0.0
# HELP zookeeper_minsessiontimeout MinSessionTimeout (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1, name2=Follower><>MinSessionTimeout)
# TYPE zookeeper_minsessiontimeout untyped
zookeeper_minsessiontimeout{memberType="Follower",replicaId="1",} 4000.0
# HELP zookeeper_numaliveconnections NumAliveConnections (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1, name2=Follower><>NumAliveConnections)
# TYPE zookeeper_numaliveconnections untyped
zookeeper_numaliveconnections{memberType="Follower",replicaId="1",} 0.0
# HELP zookeeper_lastclientresponsesize LastClientResponseSize (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1, name2=Follower><>LastClientResponseSize)
# TYPE zookeeper_lastclientresponsesize untyped
zookeeper_lastclientresponsesize{memberType="Follower",replicaId="1",} -1.0
# HELP zookeeper_maxclientresponsesize MaxClientResponseSize (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1, name2=Follower><>MaxClientResponseSize)
# TYPE zookeeper_maxclientresponsesize untyped
zookeeper_maxclientresponsesize{memberType="Follower",replicaId="1",} -1.0
# HELP zookeeper_logdirsize LogDirSize (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1, name2=Follower><>LogDirSize)
# TYPE zookeeper_logdirsize untyped
zookeeper_logdirsize{memberType="Follower",replicaId="1",} 8.05470348E8
# HELP zookeeper_packetssent PacketsSent (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1, name2=Follower><>PacketsSent)
# TYPE zookeeper_packetssent untyped
zookeeper_packetssent{memberType="Follower",replicaId="1",} 0.0
# HELP zookeeper_electiontimetaken ElectionTimeTaken (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1, name2=Follower><>ElectionTimeTaken)
# TYPE zookeeper_electiontimetaken untyped
zookeeper_electiontimetaken{memberType="Follower",replicaId="1",} 11.0
# HELP zookeeper_fsyncthresholdexceedcount FsyncThresholdExceedCount (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1, name2=Follower><>FsyncThresholdExceedCount)
# TYPE zookeeper_fsyncthresholdexceedcount untyped
zookeeper_fsyncthresholdexceedcount{memberType="Follower",replicaId="1",} 0.0
# HELP zookeeper_pendingrevalidationcount PendingRevalidationCount (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1, name2=Follower><>PendingRevalidationCount)
# TYPE zookeeper_pendingrevalidationcount untyped
zookeeper_pendingrevalidationcount{memberType="Follower",replicaId="1",} 0.0
# HELP zookeeper_txnlogelapsedsynctime TxnLogElapsedSyncTime (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1, name2=Follower><>TxnLogElapsedSyncTime)
# TYPE zookeeper_txnlogelapsedsynctime untyped
zookeeper_txnlogelapsedsynctime{memberType="Follower",replicaId="1",} -1.0
# HELP zookeeper_minclientresponsesize MinClientResponseSize (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1, name2=Follower><>MinClientResponseSize)
# TYPE zookeeper_minclientresponsesize untyped
zookeeper_minclientresponsesize{memberType="Follower",replicaId="1",} -1.0
# HELP zookeeper_ticktime TickTime (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1, name2=Follower><>TickTime)
# TYPE zookeeper_ticktime untyped
zookeeper_ticktime{memberType="Follower",replicaId="1",} 2000.0
# HELP zookeeper_maxclientcnxnsperhost MaxClientCnxnsPerHost (org.apache.ZooKeeperService<name0=ReplicatedServer_id1, name1=replica.1, name2=Follower><>MaxClientCnxnsPerHost)
# TYPE zookeeper_maxclientcnxnsperhost untyped
zookeeper_maxclientcnxnsperhost{memberType="Follower",replicaId="1",} 60.0
```
Looks like it has counters you mentioned and gauges. I see 2 options either extract counters in separate rule (I don't know how) or remove type for this rule.

Reply:
What do those potential counters look like when you graph them?

Reply:
zookeeper_packetssent and zookeeper_packetsreceived looks like a growing counters.
zookeeper_txnlogelapsedsynctime has a constant value.

Reply:
https://imgur.com/a/F6g6rdD
https://imgur.com/a/hNxHhF5
https://imgur.com/a/Wa5wvn9

Reply:
The two packet ones are counters then, we'll need to special case them.